pep-0657.rst

```diff
@@ -343,18 +343,46 @@ A reference implementation can be found in the implementation_ fork.
 Rejected Ideas
 ==============
 
-Include end line number
-^^^^^^^^^^^^^^^^^^^^^^^
-Some instructions can span across multiple lines and therefore the end offset
-and the start offset can be located in two different lines. We have decided to
-set the value for the start offset to the correct value and set a value of 0 to
-the end offset. This will result in highlighting the entire line starting from
-the value of the starting offset. The reason behind this decision is that
-storing the end line will require us to store another field similar to
-``co_lnotab``, but our traceback machinery only highlights a single line
-per frame so this information would only be used to decide to highlight to the
-end of the line. On the other hand, the end line could be useful for other
-tools such as coverage-measuring tools and tracers.
+Use a single caret instead of a range
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+It has been proposed to use a single caret instead of highlighting the full
+range when reporting errors as a way to simplify the feature. We have decided
+to not go this route for the following reasons:
+
+* Deriving the location of the caret is not straightforward using the current
+  layout of the AST. This is because the AST nodes only record the start and end
+  line numbers as well as the start and end column offsets. As the AST nodes do
+  not preserve the original tokens (by design) deriving the exact location of some
+  tokens is not possible without extra re-parsing. For instance, currently binary
+  operators have nodes for the operands but the type of the operator is stored
+  in an enumeration so its location cannot be derived from the node (this is just
+  an example of how this problem manifest, and not the only one).
+* Deriving the ranges from AST nodes greatly simplifies the implementation and reduces
+  a lot the maintenance cost and the possibilities of errors. This is because using
+  the ranges is always possible to do generically for any AST node, while any other
+  custom information would need to be extracted differently from different types of
+  nodes. Given how error-prone getting the locations manually was when this used to
+  be a manual process when generating the AST, we believe that a generic solution is
```

Comment:
> Given how error-prone getting the locations manually was when this used to
  be a manual process when generating the AST, we believe that a generic solution is
  a very important property to pursue.

There's some historical context here that I'm missing, maybe add a reference to what feature used to calculate offsets manually?

Reply:
That's the code of the old `ast.c` before 3.9

Reply:
oh I see, it is talking about the old ast generation code. Okay it makes sense in that case :)